## Applied Linear Algebra
### Amantur Amatov Homework 2 Variant 3

### 1. Find the best approximation matrix $A_1$ of rank 2 of the matrix A in the norm $||\cdot||_2$ and find $||A-A_1||_2$, where  
$$A = \begin{bmatrix}-19 & -14 & 52 & -36\\26 & 52 & -50 & 0\\-32 & -76 & -31 & 18\end{bmatrix}$$

### Solution:  

Let us find at first SVD decomposition of this matrix $A$, which equals
$$ A = V\Sigma U^* $$
After we find matrix $\Sigma$ we should reduce its rank by 1 (to get the rank 2 and $\Sigma_r$) and then get the matrix $A_1$, which equals
$$ A_1 = V\Sigma_r U^*, $$
where r = 2.

In [86]:
# Finding SVD of matrix A
import numpy as np
import sympy as sp

A = np.array([[-19, -14, 52, -36], 
              [26,52, -50, 0],
              [-32, -76, -31, 18]])
V, s, Uh = np.linalg.svd(A, full_matrices=False)
print('Matrix V:\n{}\nMatrix Sigma:\n{}\nMatrix Uh:\n{}'.format(V,np.diag(s),Uh))


Matrix V:
[[ 0.33333333 -0.66666667  0.66666667]
 [-0.66666667  0.33333333  0.66666667]
 [ 0.66666667  0.66666667  0.33333333]]
Matrix Sigma:
[[105.   0.   0.]
 [  0.  84.   0.]
 [  0.   0.  21.]]
Matrix Uh:
[[-4.28571429e-01 -8.57142857e-01  2.85714286e-01  2.77555756e-16]
 [ 6.66133815e-16 -2.85714286e-01 -8.57142857e-01  4.28571429e-01]
 [-2.85714286e-01  9.71445147e-17 -4.28571429e-01 -8.57142857e-01]]


$$ A = V \Sigma U^* = \begin{bmatrix} 1/3 & -2/3 & 2/3 \\ -2/3 & 1/3 & 2/3 \\ 2/3 & 2/3 & 1/3 \end{bmatrix} \cdot \begin{bmatrix} 105 & 0 & 0 \\ 0 & 84 & 0 \\ 0 & 0 & 21 \end{bmatrix} \cdot \begin{bmatrix} -3/7 & -6/7 & 2/7 & 0 \\ 0 & -2/7 & -6/7 & 3/7\\-2/7 & 0 & -3/7 & -6/7\end{bmatrix}$$


Now, let us reduce rank of matrix $\Sigma$.

In [87]:
# Reducing rank of matrix Sigma to rnk=2
S_rank = np.diag(s)
S_rank[2,2] = 0
S_rank

array([[105.,   0.,   0.],
       [  0.,  84.,   0.],
       [  0.,   0.,   0.]])

$$ \Sigma_{rank} = \begin{bmatrix} 105 & 0 & 0 \\ 0 & 84 & 0 \\ 0 & 0 & 0 \end{bmatrix}$$
where $rank=2$

Now, let us compute matrix $A_r$ by using matrix $\Sigma_r$.

In [88]:
# computing matrix A_1
A_r = V @ S_rank @ Uh

# checking matrix and its rank
print('Matrix A_r:\n{}\nIts rank: {}'.format(A_r,np.linalg.matrix_rank(A_r)))

Matrix A_r:
[[-15. -14.  58. -24.]
 [ 30.  52. -44.  12.]
 [-30. -76. -28.  24.]]
Its rank: 2


$$A_r = \begin{bmatrix}-15 & -13 & 58 & -24\\30 & 52 & -44 & 12\\-30 & -76 & -28 & 24\end{bmatrix}$$

So we have matrices $A$ and $A_r$. Let us find  $||A-A_r||_2$.

In [89]:
A_r = A_r.astype(int)
A = A.astype(int)
diff = A - A_r
_,s_diff,_ = np.linalg.svd(diff)
print('Second norm: {}'.format(max(s_diff)))

Second norm: 20.74161581424778


$$||A-A_r||_2 = \left\lVert\begin{bmatrix}-4 & -1 & -6 & -12\\-4 & 0 & -6 & -12\\-2 & 0 & -3 & -5\end{bmatrix}\right\rVert_2 = max_{1\leq i\leq m}{\ \sigma_i}=20.74$$

### Answers:
As we can see this norm equals approximately to the singular value of matrix $A$: $||A-A_r||_2\approx\sigma_{r+1}=21$

### 2. Estimate the relative error of the approximate solution $(1, 1)$ of the system $Ax = b$ in the norms $|\cdot|_1$ and $|\cdot|_2$ using the condition number of the matrix $A$, where
$$A = \begin{pmatrix} 1.0 & -0.06 \\-4.91 & -7.02\end{pmatrix}, b = \begin{pmatrix} 0.98 \\-12.09\end{pmatrix}$$

### Solution:
$$\hat{A} = \begin{pmatrix} 1 & 0 \\-5 & -7\end{pmatrix}, \hat{b} = \begin{pmatrix} 1 \\-12\end{pmatrix}, \hat{x} = \begin{pmatrix} 1 \\1\end{pmatrix}$$
Condition number:  
$$\chi(A) \approx \chi(\hat{A}) \approx ||\hat{A}||\cdot||\hat{A}^{-1}||
$$

So let us find at first the inverse matrix of $\hat{A}$.

In [90]:
A = np.array([[1,0],[-5,-7]])
b = np.array([[1,-12]])
x = np.array([[1,1]])
A_inv = np.linalg.inv(A)
A_inv

array([[ 1.        , -0.        ],
       [-0.71428571, -0.14285714]])

$$A^{-1} = \begin{pmatrix}1.0 & 0.0\\-0.714285714285714 & -0.142857142857143\end{pmatrix}$$

In [91]:
chi_1 = max(np.sum(np.absolute(A),axis=0))*max(np.sum(np.absolute(A_inv),axis=0))
print('Chi 1 of matrix A: ', chi_1)

Chi 1 of matrix A:  11.999999999999998


For the norm $|\cdot|_1$ we have a condition number:
$$\chi_1 = max\{6,7\} \cdot max\{1.71428571, 0.14285714\}=12$$

For the norm $|\cdot|_2$ we have a condition number:
$$\chi_2 = \sqrt{\frac{\lambda_{max}(A^*A)}{\lambda_{min}(A^*A)}}$$
Let us find $A^*A$ and then $\chi_2$.

In [92]:
AA = A.T @ A
_,singval,_ = np.linalg.svd(AA, full_matrices=True)
print('Singular Values: {}'.format(singval))
chi_2 = np.sqrt(max(singval)/min(singval))
print('Chi 2 of matrix A: ', chi_2)

Singular Values: [74.34087404  0.65912596]
Chi 2 of matrix A:  10.6201248627967


So, we have eigenvalues:  
$ \lambda_{max} = 74.34087404,\, \lambda_{min} = 0.65912596$ and $\chi_2 = 10.62$

Let us now find $\Delta b, \delta_1 b,\delta_2 b$ and $\Delta A, \delta_1 A,\delta_2 A$.

In [93]:
A_real = np.array([[1,-0.06],[-4.91,-7.02]])
b_real = np.array([[0.98,-12.09]])
Delta_b = b - b_real
Delta_A = A - A_real


delta_1_b = np.sum(np.abs(Delta_b))/np.sum(np.abs(b))
delta_2_b = np.sqrt(np.sum(Delta_b**2))/np.sqrt(np.sum(b**2))

print("""delta_1_b: {}\n
delta_2_b: {}\n""".format(delta_1_b,delta_2_b))

delta_1_b: 0.008461538461538453

delta_2_b: 0.007656414934887754



In [94]:
delta_1_A = max(np.sum(np.absolute(Delta_A),axis=0))/max(np.sum(np.absolute(A),axis=0))

dAdA = Delta_A.T @ Delta_A
_,singval_dAdA,_ = np.linalg.svd(dAdA)
delta_2_A = np.sqrt(max(singval_dAdA)/max(singval))
delta_2_A
print("""delta_1_A: {}\n
delta_2_A: {}\n""".format(delta_1_A,delta_2_A))

delta_1_A: 0.012857142857142836

delta_2_A: 0.010866415228699145



$$\Delta b = \begin{pmatrix}0.02 \\ 0.09\end{pmatrix}$$

$$\Delta A = \begin{pmatrix}0.0 & 0.06\\-0.02 & 0.02\end{pmatrix}$$

$$\delta_1 b =\frac{|\Delta b|_1}{|b|_1}\approx \frac{|\Delta b|_1}{|\hat{b}|_1} = \frac{0.11}{13}= 0.00846$$

$$\delta_2 b =\frac{|\Delta b|_2}{|b|_2}\approx \frac{|\Delta b|_2}{|\hat{b}|_2} = \frac{\sqrt{0.02^2+0.09^2}}{\sqrt{1^2+12^2}}= 0.007656$$

$$\delta_1 A =\frac{|\Delta A|_1}{|A|_1}\approx \frac{|\Delta A|_1}{|\hat{A}|_1} = \frac{0.09}{7}= 0.01286$$

For $\delta_2 A$ we find $\Delta A^* \Delta A$ and then:  
$$\delta_2 A =\frac{|\Delta A|_2}{|A|_2}\approx \frac{|\Delta A|_2}{|\hat{A}|_2} =\frac{\sigma_1 (\Delta A)}{\sigma_1 (A)} = \frac{\sqrt{0.00878}}{\sqrt{74.3408}}= 0.01087$$

Let's check the relative errors of solutions in our norms.

In [95]:
norm1_upper_bound = chi_1*(delta_1_b+delta_1_A)
norm2_upper_bound = chi_2*(delta_2_b+delta_2_A)

norm1_lower_bound = 1/chi_1*(delta_1_b+delta_1_A)
norm2_lower_bound = 1/chi_2*(delta_2_b+delta_2_A)

print("""Error 1 upper bound: {}\n
Error 1 lower bound: {}\n
Error 2 upper bound: {}\n
Error 2 lower bound: {}\n""".format(norm1_upper_bound,
                                   norm1_lower_bound,
                                   norm2_upper_bound,
                                   norm2_lower_bound))

Error 1 upper bound: 0.25582417582417544

Error 1 lower bound: 0.0017765567765567743

Error 2 upper bound: 0.1967147691496699

Error 2 lower bound: 0.001744125460188714



### Answer:
Norm $|\cdot|_2$ gave the lowest relative error $\delta_2 x \in [0.0017, 0.1967]$

### 3. Solve the system approximately and estimate the error of the solution in the norms $|\cdot|_1$, $|\cdot|_2$ and $|\cdot|_{\infty}$:
$$ \begin{cases} (-10+\varepsilon_1)x+2(-3+\varepsilon_2)y = -3+\varepsilon_3 \\ -3x + (-4+\varepsilon_1)y = -4+ \varepsilon_4 \end{cases}$$

### where the unknown numbers $\varepsilon_j$ satisfy the conditions $|\varepsilon_j| < 0.05$ for all $j=1,...,4$.

### Solution:
$$\hat{A} = \begin{pmatrix} -10 & -6 \\-3 & -4\end{pmatrix}, \hat{b} = \begin{pmatrix} -3 \\-4\end{pmatrix}, \hat{x} = \begin{pmatrix} -\frac{6}{11} \\\frac{31}{22}\end{pmatrix}$$
Condition number:  
$$\chi(A) \approx \chi(\hat{A}) \approx ||\hat{A}||\cdot||\hat{A}^{-1}||
$$

So let us find at first the inverse matrix of $\hat{A}$.

In [96]:
A = np.array([[-10,-6],[-3,-4]])
b = np.array([[-3,-4]])
x = np.array([[-6/11, 31/22]])
A_inv = np.linalg.inv(A)
A_inv

array([[-0.18181818,  0.27272727],
       [ 0.13636364, -0.45454545]])

$$A^{-1} = \begin{pmatrix}-\frac{2}{11} & \frac{3}{11}\\\frac{3}{22} & -\frac{5}{11}\end{pmatrix}$$

In [97]:
chi_1 = max(np.sum(np.absolute(A),axis=0))*max(np.sum(np.absolute(A_inv),axis=0))
print('Chi 1 of matrix A: ', chi_1)
chi_inf = max(np.sum(np.absolute(A),axis=1))*max(np.sum(np.absolute(A_inv),axis=1))
print('Chi inf of matrix A: ', chi_inf)

Chi 1 of matrix A:  9.454545454545455
Chi inf of matrix A:  9.454545454545457


For the norm $|\cdot|_1$ we have a condition number:
$$\chi_1 = max\{13,10\} \cdot max\{\frac{7}{22}, \frac{8}{11}\}=9.45$$

For the norm $|\cdot|_{\infty}$ we have a condition number:
$$\chi_{\infty} = max\{16,7\} \cdot max\{\frac{5}{11}, \frac{13}{22}\}=9.45$$

For the norm $|\cdot|_2$ we have a condition number:
$$\chi_2 = \sqrt{\frac{\lambda_{max}(A^*A)}{\lambda_{min}(A^*A}}$$
Let us find $A^*A$ and then $\chi_2$.

In [98]:
AA = A.T @ A
_,singval,_ = np.linalg.svd(AA, full_matrices=True)
print('Singular Values: {}'.format(singval))
chi_2 = np.sqrt(max(singval)/min(singval))
print('Chi 2 of matrix A: ', chi_2)

Singular Values: [157.93545699   3.06454301]
Chi 2 of matrix A:  7.178884408856446


So, we have eigenvalues:  
$ \lambda_{max} = 157.93545699,\, \lambda_{min} = 3.06454301$ and $\chi_2 = 7.18$

Let us now find $\Delta b$ and $\Delta A$.

In [99]:
e1, e2, e3, e4 = sp.symbols('varepsilon1 varepsilon2 varepsilon3 varepsilon4')
A_real = sp.Matrix([[-10+e1, 2*(-3+e2)],[-3, -4+e1]])
b_real = sp.Matrix([[-3+e3, -4+e4]])

In [100]:
Delta_A = A_real - sp.Matrix(A)
Delta_b = b_real - sp.Matrix(b)
print('Delta A equals:')
Delta_A

Delta A equals:


Matrix([
[varepsilon1, 2*varepsilon2],
[          0,   varepsilon1]])

In [101]:
print('Delta b equals:')
Delta_b
# Delta_b.subs([(e3,0.05),(e4,0.05)])

Delta b equals:


Matrix([[varepsilon3, varepsilon4]])

Because $\varepsilon_j s$ are unknown and lie in interval of $(-0.05, 0.05)$ we will evaluate the solution by the maximum interval. Let us find epsilons, where sum of deltas $A$ and $b$ in terms of norm $|\cdot|_1$ are maximum.

In [102]:
from operator import itemgetter

delta_1_Ab_list = []
for eps1 in np.arange(-0.05,0.05, 0.01).round(2):
    for eps2 in np.arange(-0.05,0.05, 0.01).round(2):
        for eps3 in np.arange(-0.05,0.05, 0.01).round(2):
            for eps4 in np.arange(-0.05,0.05, 0.01).round(2):
                Delta_b_new = Delta_b.subs([(e3,eps3),(e4,eps4)])
                Delta_b_new = np.array(Delta_b_new).astype(np.float64)
                Delta_A_new = Delta_A.subs([(e1, eps1),(e2,eps2)])
                Delta_A_new = np.array(Delta_A_new).astype(np.float64)
                
                delta_1_b = np.sum(np.abs(Delta_b_new))/np.sum(np.abs(b))
                delta_1_A = max(np.sum(np.absolute(Delta_A_new),axis=0))/max(np.sum(np.absolute(A),axis=0))
                delta_1_Ab = delta_1_A + delta_1_b
                
                delta_1_Ab_list.append((delta_1_Ab,eps1,eps2,eps3,eps4))

print(max(delta_1_Ab_list,key=itemgetter(1)))

(0.025054945054945058, 0.04, -0.05, -0.05, -0.05)


So, as we can see
$$\varepsilon_1 = 0.04$$
$$\varepsilon_2 = -0.05$$
$$\varepsilon_3 = -0.05$$
$$\varepsilon_4 = -0.05$$
Let us substitute these epsilons into $A$ and $b$.
We got:

In [103]:
Delta_A = Delta_A.subs([(e1,0.04),(e2,-0.05)])
Delta_A

Matrix([
[0.04, -0.1],
[   0, 0.04]])

In [104]:
Delta_b = Delta_b.subs([(e3,-0.05),(e4,-0.05)])
Delta_b

Matrix([[-0.05, -0.05]])

For $|\cdot|_1$:

In [105]:
delta_1_b = np.sum(np.abs(Delta_b))/np.sum(np.abs(b))
delta_1_A = max(np.sum(np.absolute(Delta_A),axis=0))/max(np.sum(np.absolute(A),axis=0))
print('delta_1_A: ', delta_1_A)
print('delta_1_b: ', delta_1_b)

delta_1_A:  0.0107692307692308
delta_1_b:  0.0142857142857143


$$\delta_1 A =\frac{|\Delta A|_1}{|A|_1}\approx \frac{|\Delta A|_1}{|\hat{A}|_1} = \frac{0.14}{13}= 0.010769$$
$$\delta_1 b =\frac{|\Delta b|_1}{|b|_1}\approx \frac{|\Delta b|_1}{|\hat{b}|_1} = \frac{0.1}{7}= 0.0142857$$

For $|\cdot|_2$:

In [106]:
Delta_b_new = np.array(Delta_b).astype(np.float64)
delta_2_b = np.sqrt(np.sum(Delta_b_new**2))/np.sqrt(np.sum(b**2))
Delta_A_new = np.array(Delta_A).astype(np.float64)

dAdA = Delta_A_new.T @ Delta_A_new
_,singval_dAdA,_ = np.linalg.svd(dAdA)
delta_2_A = np.sqrt(max(singval_dAdA)/max(singval))
print('delta_2_A: ', delta_2_A)
print('delta_2_b: ', delta_2_b)

delta_2_A:  0.009073692040505031
delta_2_b:  0.014142135623730952


For $\delta_2 A$ we find $\Delta A^* \Delta A$ and then:  

$$\delta_2 A =\frac{|\Delta A|_2}{|A|_2}\approx \frac{|\Delta A|_2}{|\hat{A}|_2} =\frac{\sigma_1 (\Delta A)}{\sigma_1 (A)} = \frac{\sqrt{0.013003}}{\sqrt{157.9354}}= 0.00907$$
$$\delta_2 b =\frac{|\Delta b|_2}{|b|_2}\approx \frac{|\Delta b|_2}{|\hat{b}|_2} = \frac{\sqrt{0.05^2+0.05^2}}{\sqrt{3^2+4^2}}= 0.014142$$

For $|\cdot|_{\infty}$:

In [107]:
delta_inf_b = np.max(np.abs(Delta_b))/np.max(np.abs(b))
delta_inf_A = max(np.sum(np.absolute(Delta_A),axis=1))/max(np.sum(np.absolute(A),axis=1))
print('delta_inf_A: ', delta_inf_A)
print('delta_inf_b: ', delta_inf_b)

delta_inf_A:  0.00875000000000000
delta_inf_b:  0.0125000000000000


$$\delta_{\infty} A =\frac{|\Delta A|_{\infty}}{|A|_{\infty}}\approx \frac{|\Delta A|_{\infty}}{|\hat{A}|_{\infty}} = \frac{0.14}{16}= 0.00875$$
$$\delta_{\infty} b =\frac{|\Delta b|_{\infty}}{|b|_{\infty}}\approx \frac{|\Delta b|_{\infty}}{|\hat{b}|_{\infty}} = \frac{0.05}{4}= 0.0125$$

Let's check the relative errors of solutions in our norms.

In [108]:
norm1_upper_bound = chi_1*(delta_1_b+delta_1_A)
norm2_upper_bound = chi_2*(delta_2_b+delta_2_A)
norm_inf_upper_bound = chi_inf*(delta_inf_b+delta_inf_A)

print("""Error 1 upper bound: {}\n
Error 2 upper bound: {}\n
Error Inf upper bound: {}""".format(norm1_upper_bound,
                                   norm2_upper_bound,
                                   norm_inf_upper_bound))

Error 1 upper bound: 0.236883116883117

Error 2 upper bound: 0.16666374325748184

Error Inf upper bound: 0.200909090909091


### Answer:
Approximate solution:
$$\hat{x} = \begin{pmatrix} -\frac{6}{11} \\\frac{31}{22}\end{pmatrix} \approx \begin{pmatrix} -0.54 \\1.41\end{pmatrix}$$
Relative errors:
$$\delta_{1} x \leq 0.236883$$
$$\delta_{2} x \leq 0.166663$$
$$\delta_{\infty} x \leq 0.200909$$

### 4. Find the approximate inverse matrix to the matrix $A$ and evaluate the approximation error with respect to the uniform norm $||\cdot||_1$, if the elements of the matrix $A$ are known with an absolute error of 0.01:
$$A \approx \begin{pmatrix}-7 & 2\\-4 & 6\end{pmatrix}$$

### Solution:
$$A \approx \begin{pmatrix}-7 & 2\\-4 & 6\end{pmatrix}$$
$$\hat{A} = \begin{pmatrix}-7 & 2\\-4 & 6\end{pmatrix}, \varepsilon = 0.01$$

The real matrix $A$ looks like
$$A = \begin{pmatrix}-7 & 2\\-4 & 6\end{pmatrix}+\begin{pmatrix}\varepsilon_{11} & \varepsilon_{12}\\\varepsilon_{21} & \varepsilon_{22}\end{pmatrix},$$
where $|\varepsilon_{ij}|\leq0.01$

Let us find inverse matrix of A.
$$\hat{A}^{-1} = \frac{1}{det \hat A}\begin{pmatrix}6 & -2\\4 & -7\end{pmatrix}=\frac{1}{-34}\begin{pmatrix}6 & -2\\4 & -7\end{pmatrix}=\,=\begin{pmatrix}-0.1765 & 0.0588\\-0.1176 & 0.2059\end{pmatrix}$$

In [109]:
# Let us check it
A_4 = np.array([[-7,2],[-4,6]])
A_4_inv = np.linalg.inv(A_4)
A_4_inv
# yep, that's it

array([[-0.17647059,  0.05882353],
       [-0.11764706,  0.20588235]])

Let us find next the conditional number $\chi$ in norm $|\cdot|_1$.  
And let us assume that $\chi (A) \approx \chi(\hat A)$

$$\chi_1(A) \approx \chi(\hat{A}) \approx ||\hat{A}||_1\cdot||\hat A^{-1}||_1=max\{11,6\}\cdot max\{0.294, 0.264\} = 11 \cdot 0.294 = 3.234
$$

So,
$$ \delta A^{-1} \leq \frac{\chi(\hat A) \delta \varepsilon}{1-\chi (\hat A)\delta \epsilon}, \delta \varepsilon = \frac {||\varepsilon||}{||\hat A||}$$

In our case with norm $|\cdot|_1$ we have:  
$$\delta \varepsilon = \frac {||\varepsilon||_1}{||\hat A||_1}=\frac {max\{0.2,0.2\}}{11}=0.0182$$
Approximation error:
$$ \delta A^{-1} \leq \frac{\chi_1(\hat A) \delta \varepsilon}{1-\chi_1 (\hat A)\delta \varepsilon}=\frac{3.234\cdot0.0182}{1-3.234*0.0182}=\frac{0.05886}{0.94114}=0.062541$$

### Answer:
$$A^{-1} \approx \begin{pmatrix}-0.1765 & 0.0588\\-0.1176 & 0.2059\end{pmatrix}$$
$$ \delta A^{-1} \leq 0.062541$$

### 5. Use simple iteration method for finding the solution of the given linear system.
$$ 
\begin{cases} 20x+4y+6z=4, \\ 3x+21y+7z=2, \\ 2x+9y+27z=5. \end{cases}
$$

### Determine the iteration number after which the approximation error for each coordinate does not exceed 0.01 and find the corresponding approximate solution. Start with $x_0 = [0, 0, 0]^T$.

### Solution:

So we have matrix A and vector b.
$$A = \begin{pmatrix}20 & 4 & 6\\3 & 21 & 7\\ 2 & 9 & 27\end{pmatrix}$$
$$B = \begin{pmatrix}4\\2\\5\end{pmatrix}$$

Let us bring our system of linear equations to the next form:
$$ x_{k+1} = P x_k+\bar b, $$
where $P=E-CA$ and $\bar b = CB$.  
We need to choose such $C$ that our iteration process converges, i.e. $||P||<1$.  
All calculations will be in terms of norm $|\cdot|_{\infty}$.

In [110]:
A = np.array([[20,4,6],[3,21,7],[2,9,27]])

B = np.array([4,2,5])

epsilon = 0.01 

Let us find this matrix $C$, matrix $P$ and vector $b$.  
$C = \begin{pmatrix}\frac{1}{20} & 0 & 0\\0 & \frac{1}{21} & 0\\ 0 & 0 & \frac{1}{27}\end{pmatrix}$  
$P = E-CA = \begin{pmatrix} 1 & 0 & 0\\0 & 1 & 0\\ 0 & 0 & 1\end{pmatrix}-\begin{pmatrix}\frac{1}{20} & 0 & 0\\0 & \frac{1}{21} & 0\\ 0 & 0 & \frac{1}{27}\end{pmatrix}\cdot\begin{pmatrix}20 & 4 & 6\\3 & 21 & 7\\ 2 & 9 & 27\end{pmatrix}$  
Its norm will be $||P||_{\infty}=0.5<1$. It means that iteration process converges.  
$b = CB = \begin{pmatrix}\frac{1}{20} & 0 & 0\\ 0 & \frac{1}{21} & 0\\ 0 & 0 & \frac{1}{27}\end{pmatrix}\cdot \begin{pmatrix}4\\2\\5\end{pmatrix}= \begin{pmatrix}0.2\\0.0952381\\0.185185\end{pmatrix}$

In [111]:
C = np.array([[1/20, 0, 0],[0, 1/21, 0],[0, 0, 1/27]])
I = np.eye(3)

P = I - np.matmul(C, A)
b = np.matmul(C,B)
# Norm_infinity of Matrix
P_norm = max(np.sum(np.absolute(P),axis=1)) # 0.5 < 1 -> converges
print('P:\n', P)
print('b:\n', b)
print('Norm_infinity of Matrix P: {}'.format(P_norm))

P:
 [[ 0.         -0.2        -0.3       ]
 [-0.14285714  0.         -0.33333333]
 [-0.07407407 -0.33333333  0.        ]]
b:
 [0.2        0.0952381  0.18518519]
Norm_infinity of Matrix P: 0.5


In purpose to find the number of iterations of the process we will use priori condition:
$$ m \geq \frac{ln(\frac{\varepsilon(1-||P||_{\infty})}
{||x^{(1)}-x^{(0)}||_{\infty}})}{ln ||P||_{\infty}}$$  
$$||x^{(1)}-x^{(0)}||_{\infty}==||b||_{\infty} = ||(0.2,\, 0.0952381,\, 0.18518519)^T||_{\infty}=0.2$$  
$$ m \geq \frac{ln(\frac{0.01*0.5)}{0.2})}{ln 0.5}=5.32$$
Let us do the iteration process with number of iterations of 6 using this formula:
$$ x_{k+1} = P x_k+\bar b$$

In [112]:
# Initial and next values of x 
x_0 = np.array([0,0,0])
x_1 = np.matmul(P, x_0) + b

# Norm_infinity of Vector
norm_diff = np.max(np.abs(x_1 - x_0))
#apriori condition
m = np.log(epsilon*(1-P_norm)/norm_diff)/np.log(P_norm)
print(m)
print('Number of iterations: {:.0f}'.format(m+1))

#loop with number of iterations = 6
x = x_0
for i in range(int(m)+1):
    x = np.matmul(P, x) + b
    print(x)
#     print('$$x_{}={}^T$$'.format(i+1, sp.latex(sp.Matrix(x.round(4)).T)))
    
print('\nSolution: {}'.format(x))

5.321928094887363
Number of iterations: 6
[0.2        0.0952381  0.18518519]
[0.12539683 0.00493827 0.13862434]
[0.15742504 0.03111615 0.17425044]
[0.14150164 0.01466532 0.16315202]
[0.14812133 0.02063957 0.16981514]
[0.14492754 0.01747286 0.16733338]

Solution: [0.14492754 0.01747286 0.16733338]


$$x_1=\left[\begin{matrix}0.2 & 0.0952 & 0.1852\end{matrix}\right]^T$$
$$x_2=\left[\begin{matrix}0.1254 & 0.0049 & 0.1386\end{matrix}\right]^T$$
$$x_3=\left[\begin{matrix}0.1574 & 0.0311 & 0.1743\end{matrix}\right]^T$$
$$x_4=\left[\begin{matrix}0.1415 & 0.0147 & 0.1632\end{matrix}\right]^T$$
$$x_5=\left[\begin{matrix}0.1481 & 0.0206 & 0.1698\end{matrix}\right]^T$$
$$x_6=\left[\begin{matrix}0.1449 & 0.0175 & 0.1673\end{matrix}\right]^T$$

### Answer:
$$x_{ans}=\left[\begin{matrix}0.1449 & 0.0175 & 0.1673\end{matrix}\right]^T$$

### 6. Find the most influential vertex in the graph:
$$A = \left[\begin{matrix}1 & 1 & 1 & 0 & 0\\1 & 0 & 0 & 0 & 1\\1 & 1 & 0 & 0 & 1\\1 & 0 & 1 & 1 & 1\\1 & 0 & 0 & 0 & 0\end{matrix}\right]$$
### using the PageRank algorithm with $β = 0.15$, where the graph adjacency matrix is defined as follows

In [113]:
A = np.array([[1, 1, 1, 0, 0],
              [1, 0, 0, 0, 1],
              [1, 1, 0, 0, 1],
              [1, 0, 1, 1, 1],
              [1, 0, 0, 0, 0]])

Let us build matrix of possibilities $P$.

In [114]:
P = np.array([A[0]/3, A[1]/2, A[2]/3, A[3]/4, A[4]/1])
sp.Matrix(P.round(3))

Matrix([
[0.333, 0.333, 0.333,  0.0,   0.0],
[  0.5,   0.0,   0.0,  0.0,   0.5],
[0.333, 0.333,   0.0,  0.0, 0.333],
[ 0.25,   0.0,  0.25, 0.25,  0.25],
[  1.0,   0.0,   0.0,  0.0,   0.0]])

Now, let us build matrix $Q$ in order to find matrix $\hat P$.  

$$\hat P = P\cdot(1-\beta)+Q\cdot\beta,$$
where $\beta = 0.15$

In [115]:
n = 5

Q = np.ones((5,5))/n

beta = 0.15

P_hat = P * (1 - beta) + Q * beta

sp.Matrix(P_hat)

Matrix([
[0.313333333333333, 0.313333333333333, 0.313333333333333,   0.03,              0.03],
[            0.455,              0.03,              0.03,   0.03,             0.455],
[0.313333333333333, 0.313333333333333,              0.03,   0.03, 0.313333333333333],
[           0.2425,              0.03,            0.2425, 0.2425,            0.2425],
[             0.88,              0.03,              0.03,   0.03,              0.03]])

We will begin with $X_0$ with equal possibilities.
$$X_0 = \left[\begin{matrix}0.2\\0.2\\0.2\\0.2\\0.2\end{matrix}\right]$$
Iteration will process until the difference between $x_{i+1}$ and $x_{i}$ will be small enough ($\varepsilon=0.001$).

In [116]:
x = np.ones(5)/n
cond = True
i = 1
while cond:
    x_new = np.matmul(P_hat.T, x)
    print(x_new.round(4))
#     print('$$x_{}={}^T$$'.format(i,sp.latex(sp.Matrix(x_new.round(4)).T)))
    diff = x_new - x
    if (diff < 0.001).all():
        cond = False
    x = x_new
    i += 1

[0.4408 0.1433 0.1292 0.0725 0.2142]
[0.4499 0.1915 0.1703 0.0454 0.1429]
[0.4182 0.2057 0.1671 0.0396 0.1693]
[0.4356 0.1958 0.1569 0.0384 0.1732]
[0.4365 0.1979 0.1616 0.0382 0.1659]
[0.4327 0.1995 0.1618 0.0381 0.168 ]
[0.4341 0.1984 0.1607 0.0381 0.1687]
[0.4343 0.1985 0.1611 0.0381 0.168 ]


$$x_1=\left[\begin{matrix}0.4408 & 0.1433 & 0.1292 & 0.0725 & 0.2142\end{matrix}\right]^T$$

$$x_2=\left[\begin{matrix}0.4499 & 0.1915 & 0.1703 & 0.0454 & 0.1429\end{matrix}\right]^T$$

$$x_3=\left[\begin{matrix}0.4182 & 0.2057 & 0.1671 & 0.0396 & 0.1693\end{matrix}\right]^T$$

$$x_4=\left[\begin{matrix}0.4356 & 0.1958 & 0.1569 & 0.0384 & 0.1732\end{matrix}\right]^T$$

$$x_5=\left[\begin{matrix}0.4365 & 0.1979 & 0.1616 & 0.0382 & 0.1659\end{matrix}\right]^T$$

$$x_6=\left[\begin{matrix}0.4327 & 0.1995 & 0.1618 & 0.0381 & 0.168\end{matrix}\right]^T$$

$$x_7=\left[\begin{matrix}0.4341 & 0.1984 & 0.1607 & 0.0381 & 0.1687\end{matrix}\right]^T$$

$$x_8=\left[\begin{matrix}0.4343 & 0.1985 & 0.1611 & 0.0381 & 0.168\end{matrix}\right]^T$$

### Answer:
So the most infuential vertex in this Graph is vertex $1$ with possibility $P=0.4343$.

### 7. Find the value $f(A)$ of the function $f(l) = ln(l + 4)$, where
$$A = \left[\begin{matrix}9 & -5 & -6\\-21 & 16 & 17\\24 & -15 & -17\end{matrix}\right]$$

### Solution:
Because our function looks like $f(l) = ln(l + 4)$ we can calculate $\hat A = A + I*4$.

In [117]:
A = np.array([[9, -5, -6],
              [-21, 16, 17],
              [24, -15, -17]])
A = A + np.eye(3)*4

A_hat = sp.Matrix(A)
A_hat

Matrix([
[ 13.0,  -5.0,  -6.0],
[-21.0,  20.0,  17.0],
[ 24.0, -15.0, -13.0]])

Now our task is to find $f(\hat A)$.  
As we know, $$\hat A = T\cdot J \cdot T_T,$$
where $T$ is a transform matrix which consists of eigenvectors of matrix $\hat A$ and $J$ is a Jordan matrix, which consists of eigenvalues. So, at first let us find eigenvalues of our matrix.

In [118]:
lmd = sp.symbols('lambda')
A_hat_lmd = A_hat-sp.eye(3)*lmd
A_hat_lmd

Matrix([
[13.0 - lambda,          -5.0,           -6.0],
[        -21.0, 20.0 - lambda,           17.0],
[         24.0,         -15.0, -lambda - 13.0]])

$$\lambda^{3} + 20.0 \lambda^{2} - 125.0 \lambda + 250.0 = (\lambda-5)^2(\lambda-10)=0$$

In [119]:
sp.solve(A_hat_lmd.det())

[5.00000000000000, 10.0000000000000]

So, after solving this cubic equations we can build Jordan Matrix.
$$J = \left[\begin{matrix}5.0 & 1.0 & 0\\0 & 5.0 & 0\\0 & 0 & 10.0\end{matrix}\right]$$

Now, let us find eigenvectors of this matrix.
First vector of eigenvalue 5:
$$(A-I\cdot5)v=0$$

In [120]:
from sympy import *
x1, x2, x3 = sp.symbols('v1, v2, v3')
b = sp.Matrix([0, 0, 0])

# first lambda = 5 first vector
A_hat_5_1 = A_hat-sp.eye(3)*5
A_hat_5_1
system = (A_hat_5_1, b)
linsolve(system, (x1, x2, x3))

FiniteSet((0.333333333333333*v3, -0.666666666666667*v3, v3))

So, let us take $v_3 = 9$. 
$$v_1=(3,-6,9)^T$$

Let us find generalized second eigenvector for eigenvalue 5, because its degree equals 2, such that $v_2\neq0$.
$$(A-I\cdot5)^2v_2=0$$

In [121]:
# first lambda = 5 second vector
A_hat_5_2 = (A_hat-sp.eye(3)*5)**2
A_hat_5_2
system = (A_hat_5_2, b)
linsolve(system, (x1, x2, x3))

FiniteSet((1.0*v2 + 1.0*v3, v2, v3))

So, let us take $v_2 = 1$ and $v_3 = 0$. 
$$v_2=(1,1,0)^T$$

Let us find eigenvector for eigenvalue 10.
$$(A-I\cdot10)v_3=0$$

In [122]:
# second lambda = 10
A_hat_10 = A_hat-sp.eye(3)*10
system = (A_hat_10, b)
linsolve(system, (x1, x2, x3))

FiniteSet((0.333333333333333*v3, -1.0*v3, v3))

So, let us take $v_3 = 1$. 
$$v_2=(\frac{1}{3},-1,1)^T$$

Now we can build transform matrix from these eigenvectors.

In [123]:
T = sp.Matrix([[3,1,1/3],[-6,1,-1],[9,0,1]])
T

Matrix([
[ 3, 1, 0.333333333333333],
[-6, 1,                -1],
[ 9, 0,                 1]])

In [124]:
# check with library 
T, J = A_hat.jordan_form()
T

Matrix([
[ 3.0, 1.0, 0.333333333333333],
[-6.0, 1.0,              -1.0],
[ 9.0,   0,               1.0]])

So, as we have found Jordan matrix and transform matrix, we can use this formula for estimating $log(A)$.
$$f(A)=T\cdot f(J) T^T,$$
where $$f(J)=\left[\begin{matrix}\frac{f(\lambda_1)}{0!} &  \frac{f'(\lambda_1)}{1!} & 0\\0 & \frac{f(\lambda_1)}{0!}) & 0\\0 & 0 & f\frac{f(\lambda_2)}{0!})\end{matrix}\right]=\left[\begin{matrix} log(5) &  \frac{1}{5} & 0\\0 & log(5) & 0\\0 & 0 & log(10)\end{matrix}\right]$$

I wrote a function, which computes the degree of derivative of log(x).   

Its input: substitution value and degree of derivative.

In [125]:
def ln_der(subs, deg):
    x = sp.Symbol('x')
    degn = 0
    expr = lambda x: sp.ln(x)
    diff = lambda f: sp.diff(f, x)
    
    if deg == 0:
        func = expr(x)
        f = sp.lambdify(x, expr(subs))
        return f(subs)#, func 
    
    while degn <= deg:
        if degn == 0:
            func = diff(expr(x))
        else:
            func = diff(func)
        f = sp.lambdify(x, func/sp.factorial(deg))
        degn += 1
        
    return f(subs)#, func/sp.factorial(deg)

In [126]:
f_J = np.array([[ln_der(J[0],0),ln_der(J[0],1),0],
                [0,ln_der(J[4],0),0],
                [0,0,ln_der(J[8],0)]])
sp.Matrix(f_J)

Matrix([
[1.6094379124341,           -0.04,                0],
[              0, 1.6094379124341,                0],
[              0,               0, 2.30258509299405]])

As we found $f(J)$, now we can find $f(A)=T\cdot f(J)\cdot T^T$

In [127]:
f_A = (T * f_J * T.inv()).evalf(4)
# print(sp.latex(f_A))

### Answer:
$$f(A) = \left[\begin{matrix}2.183 & -0.6931 & -0.6531\\-1.839 & 3.689 & 1.999\\1.719 & -2.079 & -0.35\end{matrix}\right]$$